tests/test_timer.c

```diff
@@ -0,0 +1,248 @@
+
+	timer_link_t *t2 = new_timer(TICKS_FROM_MS(2), TF_CONTINUOUS, cb2);
+	DEFER(delete_timer(t2));
+
+	timer_link_t *t3 = new_timer(TICKS_FROM_MS(7), TF_CONTINUOUS, cb3);
+	DEFER(delete_timer(t3));
+
+	timer_link_t *t1 = new_timer(TICKS_FROM_MS(11), TF_ONE_SHOT, cb1);
+	DEFER(delete_timer(t1));
+
+	uint64_t tick0 = timer_ticks();
+
+	wait_ms(12);
+	stop_timer(t2);
+
+	wait_ms(20);
+	stop_timer(t3);
+
+	uint8_t expected[] = { 2, 2, 2, 3, 2, 2, 1, 2, 3, 3, 3, 0 };
+	ASSERT_EQUAL_MEM((uint8_t*)called_list, expected, sizeof(expected), "invalid order of timer callbacks");
+
+	// Check that timer_ticks return approximate correct value across all timer executions
+	tick0 = timer_ticks() - tick0;
+	ASSERT_EQUAL_SIGNED(
+		TIMER_MICROS(tick0) / 1000, 
+		12+20,
+		"invalid timer_ticks");
+}
+
+void test_timer_slow_callback(TestContext *ctx) {
+	timer_init();
+	DEFER(timer_close());
+
+	// Check that if a callback is too slow, it doesn't prevent other
+	// timers from running.
+	volatile int called_slow = 0;
+	volatile int called_fast = 0;
+	void slow(int ovlf) { 
+		wait_ms(10);
+		called_slow++;
+	}
+
+	timer_link_t *t4;
+	void fast(int ovlf) {
+		called_fast++;
+	}
+
+	timer_link_t *t1 = new_timer(TICKS_FROM_MS(4), TF_ONE_SHOT, slow);
+	DEFER(delete_timer(t1));
+	timer_link_t *t2 = new_timer(TICKS_FROM_MS(2), TF_ONE_SHOT, slow);
+	DEFER(delete_timer(t2));
+	timer_link_t *t3 = new_timer(TICKS_FROM_MS(5), TF_ONE_SHOT, slow);
+	DEFER(delete_timer(t3));
+	t4 = new_timer(TICKS_FROM_MS(2), TF_CONTINUOUS, fast);
+	DEFER(delete_timer(t4));
+
+	uint64_t tick0 = timer_ticks();
+	wait_ms(10);
+	uint64_t tick1 = timer_ticks();
+
+	ASSERT_EQUAL_SIGNED(called_slow, 3, "slow timers not called");
+
+	// The total execution time is 30ms (3 slow timers) + 2ms (time before
+	// the first slow timer fires). The fast timer is run every 2ms in this
+	// interval.
+	ASSERT_EQUAL_SIGNED(called_fast, (30+2)/2, "fast timers not called");
+
+	ASSERT_EQUAL_SIGNED(
+		TIMER_MICROS(tick1-tick0) / 1000, 
+		30+2,
+		"invalid timer_ticks");
+}
+
+// Change the hardware count register.
+static uint32_t write_count(uint32_t x) {
+	extern uint32_t last_interrupt_tick;
+
+	uint32_t old = TICKS_READ();
+
+	asm volatile("mtc0 %0,$9\n\t nop \n\t" :  : "r" (x) );
+
+	// Since we're jumping around in time, we need to also update
+	// "last_interrupt_tick", which is required by timer_ticks() when
+	// interrupts are disabled to provide the 64-bit counter. The idea
+	// of this variable is to contain a "recent snapshot" of the counter.
+	// So we just put x-1 there.
+	last_interrupt_tick = x - 1;
+
+	return old;
+}
+
+void test_timer_ticks(TestContext *ctx) {
+	timer_init();
+	DEFER(timer_close());
+
+	// We want to fuzz different conditions around the overflow, that is
+	// when the hardware counter goes to zero. That's where all problems lie
+	// in the implementation. Try first to be exhaustive. Also, check with
+	// both enabled and disabled interrupts because the codepaths are different.
```

Comment:
I hope the explanation above clarifies. I for sure am testing that condition because it happens any time there's an overflow while interrupts are disabled, and the test is executed with and without interrupts for this very reason.